### ISSUE TITLE
Prevent overlap of controllers

### ISSUE THREAD

**Issue by chrisgoringe** [chrisgoringe/cg-controller#336]

**Comment by JorgeR81** [comment]
I think once this is solved, this other issue would be mostly solved ( https://github.com/chrisgoringe/cg-controller/issues/345 )

---

The main problem is when we wan't so snap 2 clusters of windows. Like this:

![z1](https://github.com/user-attachments/assets/b8255ed9-f24c-4378-8ca0-74688015dcab)

---

Snapping a cluster of windows works, if only 2 windows make contact.  
So here, snapping to the right works fine. The vertical gutters are perfectly aligned. 


![z2](https://github.com/user-attachments/assets/55d44cfb-b288-4b9e-90e2-6f230ed8bd57)

---

Sampling a single controller to the right and to the top also works perfectly.  

![z3](https://github.com/user-attachments/assets/d5c588ae-df76-4704-9ccc-175b5d1ef9c7)

---

The problem is when multiple windows need to make contact at the same time.
In this case, the windows will not align ( to the top ) automatically.

So unless we are very precise while moving, the Controllers may overlay a little. 
If we prevent Controller overlay, we could also prevent this from happening.

![z4](https://github.com/user-attachments/assets/ecaaf0f6-82a1-4835-be2b-15d78f32c4ed)

---

**This is more than a small aesthetic issue.**

It could break the whole snapping configuration.

For instance, it could result in snapping 2 Controllers that shouldn't be snapped.
But now they are snapped and they will move together. Like this:

![z5](https://github.com/user-attachments/assets/aaa96a20-ec5f-492b-8573-75c7572c01da)

**Comment by chrisgoringe** [comment]
I don't think this is actually about overlap. If controllers overlap by less than the snapping distance they should already snap *outwards* - the snapping distance works in both directions (try a simple case, overlapping two controllers by a few pixels - they snap to a neat fit.

The problem you are having here is to do with the sequence in which the different snapping rules get applied.

Here's a really simple example:
![Screenshot 2024-11-18 115220](https://github.com/user-attachments/assets/5fa5ffaf-3a68-4a76-b0c2-3f145b1614ae)

the two lower controllers are close enough to the one above that they should snap to it. But what's happening is that one controller is snapping up, but then being pulled back down by the rule that keeps its top aligned with the top of the other one.

**Comment by chrisgoringe** [comment]
I think I've found a way to improve it in quite a lot of scenarios. See how it works now.

**Comment by JorgeR81** [comment]
> I think I've found a way to improve it in quite a lot of scenarios. See how it works now.

Yes, it works perfectly for the case I described above !
